hermes-frontend/src/main/java/pl/allegro/tech/hermes/frontend/server/TopicMetadataLoadingRunner.java

```diff
@@ -0,0 +1,98 @@
+import pl.allegro.tech.hermes.frontend.metric.CachedTopic;
+import pl.allegro.tech.hermes.frontend.producer.BrokerMessageProducer;
+
+import javax.inject.Inject;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import static java.util.stream.Collectors.toList;
+import static pl.allegro.tech.hermes.common.config.Configs.FRONTEND_STARTUP_TOPIC_METADATA_LOADING_THREAD_POOL_SIZE;
+import static pl.allegro.tech.hermes.frontend.server.CompletableFuturesHelper.allComplete;
+import static pl.allegro.tech.hermes.frontend.server.MetadataLoadingResult.Type.FAILURE;
+import static pl.allegro.tech.hermes.frontend.server.MetadataLoadingResult.Type.SUCCESS;
+
+public class TopicMetadataLoadingRunner {
+
+    private static final Logger logger = LoggerFactory.getLogger(TopicMetadataLoadingRunner.class);
+
+    private final BrokerMessageProducer brokerMessageProducer;
+
+    private final TopicsCache topicsCache;
+
+    private final int retryCount;
+
+    private final long retryInterval;
+
+    private final int threadPoolSize;
+
+    @Inject
+    public TopicMetadataLoadingRunner(BrokerMessageProducer brokerMessageProducer,
+                                      TopicsCache topicsCache,
+                                      ConfigFactory config) {
+        this(brokerMessageProducer,
+                topicsCache,
+                config.getIntProperty(Configs.FRONTEND_STARTUP_TOPIC_METADATA_LOADING_RETRY_COUNT),
+                config.getLongProperty(Configs.FRONTEND_STARTUP_TOPIC_METADATA_LOADING_RETRY_INTERVAL),
+                config.getIntProperty(FRONTEND_STARTUP_TOPIC_METADATA_LOADING_THREAD_POOL_SIZE));
+    }
+
+    TopicMetadataLoadingRunner(BrokerMessageProducer brokerMessageProducer,
+                               TopicsCache topicsCache,
+                               int retryCount,
+                               long retryInterval,
+                               int threadPoolSize) {
+        this.brokerMessageProducer = brokerMessageProducer;
+        this.topicsCache = topicsCache;
+        this.retryCount = retryCount;
+        this.retryInterval = retryInterval;
+        this.threadPoolSize = threadPoolSize;
+    }
+
+    public void refreshMetadata() {
+        long start = System.currentTimeMillis();
+        logger.info("Loading topics metadata");
+        List<CachedTopic> topics = topicsCache.getTopics();
+        List<MetadataLoadingResult> allResults = loadMetadataForTopics(topics);
+        logResultInfo(allResults, System.currentTimeMillis() - start);
+    }
+
+    private List<MetadataLoadingResult> loadMetadataForTopics(List<CachedTopic> topics) {
+        try (TopicMetadataLoader loader = new TopicMetadataLoader(brokerMessageProducer, retryCount, retryInterval, threadPoolSize)) {
+            return allComplete(topics.stream().map(loader::loadTopicMetadata).collect(toList())).join();
```

Comment:
I have an idea for optimization. Instead of refreshing metadata for all topics maybe we should only refresh metadata for unused topics. Unused topic can be a topic for which `topic.meter.getOneMinuteRate() == 0`. Thanks to this much less requests will be sent to Kafka. What do you say about this?

Reply:
Ok, we’ve discussed this and I learned that kafka-producer doesn’t send metadata request if it has required information. Therefore there is no need for further optimization.